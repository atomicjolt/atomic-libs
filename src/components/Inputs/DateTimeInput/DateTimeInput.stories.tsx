import { ComponentMeta, ComponentStory } from "@storybook/react";
import DateTimeInput, { Props } from ".";
import { DefaultInputProperties } from "../../../utils";

export default {
  title: "Inputs/DateTimeInput",
  component: DateTimeInput,
} as ComponentMeta<typeof DateTimeInput>;

const Template: ComponentStory<typeof DateTimeInput> = (args: Props<any>) => (
  <DateTimeInput {...args} />
);

export const Default = Template.bind({});
Default.args = {
  value: new Date("2022-02-21"),
  min: new Date("2022-01-01"),
  max: new Date("2023-01-01"),
  ...DefaultInputProperties,
  label: "Date input label",
  message: "MM/DD/YYYY, HH:MM AM/PM",
  readonly: false,
};

// Additional stories go here
