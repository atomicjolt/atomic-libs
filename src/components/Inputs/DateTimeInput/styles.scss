@use "../../defines" as *;

.aj-input input[type="datetime-local"] {
  --input-width-md: 260px;
  --input-width-lg: 360px;
}
