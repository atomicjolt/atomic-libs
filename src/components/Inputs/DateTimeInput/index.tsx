import React, { useMemo } from "react";
import cn from "classnames";
import { useIds, useIsoDateString } from "../../../hooks";
import { SharedInputProps } from "../../../types";
import "../../general.scss";
import InputLabel from "../../shared/InputLabel";
import "../common.scss";
import "./styles.scss";

export interface Props<T> extends SharedInputProps {
  value: T;
  /** Whatever type the value is provided as, the new value will be provided as.
   * If value is a string, `onChange` will recieve a string, if it is a Date,
   * `onChange` will recieve a Date
   */
  onChange: (value: T) => void;
  min?: string | Date | number;
  max?: string | Date | number;
  readonly?: boolean;
}

/** Date Input Component */
export default function DateTimeInput<T extends string | Date | number>({
  value: initialValue,
  min: initialMin,
  max: initialMax,
  onChange,
  label,
  error,
  message,
  size = "medium",
  readonly = false,
  disabled = false,
  required = false,
  hideLabel = false,
}: Props<T>) {
  const [inputId, errorId] = useIds("date-input", ["input", "error"]);

  const value = useIsoDateString(initialValue);
  const min = useIsoDateString(initialMin);
  const max = useIsoDateString(initialMax);

  return (
    <div
      className={cn("aj-input", `is-${size}`, {
        "has-error": error,
        "is-disabled": disabled,
      })}
    >
      <InputLabel message={message} htmlFor={inputId} hidden={hideLabel}>
        {label}
      </InputLabel>
      <div className="aj-input__date">
        <input
          id={inputId}
          aria-describedby={error ? errorId : ""}
          type="datetime-local"
          min={min}
          max={max}
          value={value}
          readOnly={readonly}
          disabled={disabled}
          required={required}
          onChange={(e) => {
            if (typeof initialValue === "string") {
              onChange(e.target.value as T);
            } else if (typeof initialValue === "number") {
              onChange(new Date(e.target.value).getTime() as T);
            } else {
              onChange(new Date(e.target.value) as T);
            }
          }}
        />
      </div>
      {error && (
        <p id={errorId} className="aj-label--error">
          {error}
        </p>
      )}
    </div>
  );
}
