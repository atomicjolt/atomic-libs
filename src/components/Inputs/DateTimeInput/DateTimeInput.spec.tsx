import { render } from "@testing-library/react";
import DateTimeInput from ".";

it("matches snapshot", () => {
  const result = render(
    <DateTimeInput
      label="Text Label"
      message="message"
      error="some error"
      value={new Date("2022-02-18")}
      onChange={(value) => {}}
    />
  );
  expect(result.asFragment()).toMatchSnapshot();
});
