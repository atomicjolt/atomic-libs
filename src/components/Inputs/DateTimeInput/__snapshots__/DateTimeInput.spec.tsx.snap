// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="aj-input is-medium has-error"
  >
    <label
      class="aj-label"
      for="date-input-input-12345678900000000000"
    >
      Text Label
      <p
        class="aj-label--message"
      >
        message
      </p>
    </label>
    <div
      class="aj-input__date"
    >
      <input
        aria-describedby="date-input-error-12345678900000000000"
        id="date-input-input-12345678900000000000"
        max=""
        min=""
        type="datetime-local"
        value="2022-02-18T00:00"
      />
    </div>
    <p
      class="aj-label--error"
      id="date-input-error-12345678900000000000"
    >
      some error
    </p>
  </div>
</DocumentFragment>
`;
